Implement text-representation for CounterStyle

vitorroriz opened the pull request:
#### 6aaf9c99077bb6f1cc0856d5380a5bca6f6cfeaf
<pre>
Implement text-representation for CounterStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=249804">https://bugs.webkit.org/show_bug.cgi?id=249804</a>
rdar://103648354

Reviewed by Tim Nguyen.

Implementing the text representation for the integer counter value
for a defined counter-style.

The text representation is specified at <a href="https://www.w3.org/TR/css-counter-styles-3/#counter-styles">https://www.w3.org/TR/css-counter-styles-3/#counter-styles</a>

A key point of the text representation is the algorithm used, which is
defined by the counter-style system. Each algorithm
is defined at <a href="https://www.w3.org/TR/css-counter-styles-3/#counter-style-system">https://www.w3.org/TR/css-counter-styles-3/#counter-style-system</a>

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemCyclic):
(WebCore::CSSCounterStyle::counterForSystemFixed):
(WebCore::CSSCounterStyle::counterForSystemSymbolic):
(WebCore::CSSCounterStyle::counterForSystemAlphabetic):
(WebCore::CSSCounterStyle::counterForSystemNumeric):
(WebCore::CSSCounterStyle::counterForSystemAdditive):
(WebCore::CSSCounterStyle::initialRepresentation):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::shouldApplyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
(WebCore::CSSCounterStyle::createCounterStyleDecimal):
* Source/WebCore/css/CSSCounterStyle.h:
(WebCore::CSSCounterStyle::fallback const):
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::decimalCounter):
(WebCore::CSSCounterStyleRegistry::decimalCounter const): Deleted.
* Source/WebCore/css/CSSCounterStyleRegistry.h:

Canonical link: <a href="https://commits.webkit.org/260792@main">https://commits.webkit.org/260792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f55e4e0e5ca1d09a7e262594cfcd65622849cd22

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18568 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42206 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/942 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118608 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113329 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9765 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101720 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115200 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98150 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43126 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29803 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84883 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11312 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31146 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11976 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8081 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17342 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50750 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13707 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    if (!amountOfSymbols || !value) {
```
`value <= 0` instead of `!value`?

Also, can extends counter-style rules define symbolic rules without symbols? Is it correct to assert in that case?

vitorroriz replied:
if author uses extends system, it can't define symbols, as defined by the spec. It should inherit the symbols from whatever it is extending. If no symbols are defined, the parser should invalidate the rule. Some systems even require at least 2 symbols, for example.

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+    if (!value)
```
I would have said to check explicitly for 0, but I remembered that the webkit style guide forbids that :P

vitorroriz replied:
I personally would also prefer it!

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
+                return symb;
+        }
+        return { };
+    }
```
can we avoid abbreviations? `additiveSymbols`/`symbol`/`weight`, same below

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
+            continue;
+        auto repetitions = static_cast<unsigned>(std::floor(value / weight));
+        appendToResult(symb, repetitions);
+        value -=  weight * repetitions;
```
nit: extra space after `=`

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
+
+String CSSCounterStyle::fallbackText(int value)
+{
+    if (m_isFallingBack || !fallback().get()) {
+        m_isFallingBack = false;
+        return CSSCounterStyleRegistry::decimalCounter()->text(value);
+    }
+    m_isFallingBack = true;
+    auto fallbackText = fallback()->text(value);
+    m_isFallingBack = false;
+    return fallbackText;
```
Is this `m_isFallingBack` field useful? Doesn't `fallback()->text()` eventually call `fallbackText()` again anyway?

vitorroriz replied:
Yes, because we need to detect loops and fallback to decimal in that case

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
+bool CSSCounterStyle::shouldApplyNegativeSymbols(int value) const
+{
+    auto sys = system();
+    return value < 0 && (sys == CSSCounterStyleDescriptors::System::Symbolic || sys == CSSCounterStyleDescriptors::System::Numeric || sys == CSSCounterStyleDescriptors::System::Alphabetic || sys == CSSCounterStyleDescriptors::System::Additive);
```
Can we avoid the abbreviation? `auto system = this->system();`

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -27,52 +27,211 @@
 
-String CSSCounterStyle::counterForSystemCyclic(int)
+// https://www.w3.org/TR/css-counter-styles-3/#cyclic-system
+String CSSCounterStyle::counterForSystemCyclic(int value)
```
I forgot to mention, but can we make all of these counterFor* functions `const`?

vitorroriz committed:
Implement text-representation for CounterStyle
https://bugs.webkit.org/show_bug.cgi?id=249804
rdar://103648354

Reviewed by Tim Nguyen.

Implementing the text representation for the integer counter value
for a defined counter-style.

The text representation is specified at https://www.w3.org/TR/css-counter-styles-3/#counter-styles

A key point of the text representation is the algorithm used, which is
defined by the counter-style system. Each algorithm
is defined at https://www.w3.org/TR/css-counter-styles-3/#counter-style-system

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemCyclic):
(WebCore::CSSCounterStyle::counterForSystemFixed):
(WebCore::CSSCounterStyle::counterForSystemSymbolic):
(WebCore::CSSCounterStyle::counterForSystemAlphabetic):
(WebCore::CSSCounterStyle::counterForSystemNumeric):
(WebCore::CSSCounterStyle::counterForSystemAdditive):
(WebCore::CSSCounterStyle::initialRepresentation):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::shouldApplyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
(WebCore::CSSCounterStyle::createCounterStyleDecimal):
* Source/WebCore/css/CSSCounterStyle.h:
(WebCore::CSSCounterStyle::fallback const):
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::decimalCounter):
(WebCore::CSSCounterStyleRegistry::decimalCounter const): Deleted.
* Source/WebCore/css/CSSCounterStyleRegistry.h:

Canonical link: https://commits.webkit.org/260792@main
